gnd/src/main/java/com/google/android/gnd/persistence/remote/firestore/FirestoreStorageManager.java

```diff
@@ -65,27 +68,35 @@ private StorageReference createReference(String path) {
     return createReference(remoteDestinationPath).getDownloadUrl();
   }
 
-  /** Upload file to Firebase Storage. */
   @Override
-  public Completable uploadMediaFromFile(File file, String remoteDestinationPath) {
-    return Completable.create(
+  public Flowable<Event<UploadProgress>> uploadMediaFromFile(
+      File file, String remoteDestinationPath) {
+    return Flowable.create(
         emitter ->
             createReference(remoteDestinationPath)
                 .putFile(Uri.fromFile(file))
                 .addOnCompleteListener(uploadTask -> emitter.onComplete())
-                .addOnFailureListener(emitter::onError)
+                .addOnPausedListener(
+                    taskSnapshot -> {
+                      emitter.onNext(Event.create(new UploadProgress(UploadState.PAUSED)));
+                    })
+                .addOnFailureListener(
+                    throwable -> {
+                      emitter.onNext(Event.create(new UploadProgress(UploadState.FAILED)));
+                      emitter.onError(throwable);
+                    })
                 .addOnSuccessListener(
                     taskSnapshot -> {
-                      // TODO: taskSnapshot contains metadata that can be displayed after uploading
+                      emitter.onNext(Event.create(new UploadProgress(UploadState.COMPLETED)));
                     })
                 .addOnProgressListener(
-                    taskSnapshot -> {
-                      // TODO: Display upload status in app or notification
-                      double completed =
-                          100.0
-                              * taskSnapshot.getBytesTransferred()
-                              / taskSnapshot.getTotalByteCount();
-                      Timber.d("Uploading in progress: %s %f", remoteDestinationPath, completed);
-                    }));
+                    taskSnapshot ->
+                        emitter.onNext(
+                            Event.create(
+                                new UploadProgress(
+                                    UploadState.IN_PROGRESS,
+                                    (int) taskSnapshot.getTotalByteCount(),
+                                    (int) taskSnapshot.getBytesTransferred())))),
+        BackpressureStrategy.BUFFER);
```

Comment:
Only use buffer if it's important that downstream subscribers catch every single value. In this case, only the latest value is relevant, so we'd use LATEST.